### PR TITLE
Fix HeaderCarrier changing header casing

### PR DESCRIPTION
**Pull request by robertmarsal**
At the moment the `HeaderCarrier` will normalise HTTP headers and capitalise them. This way `traceparent` becomes `Traceparent` and is sent as an uppercase header. This is against the [W3C recommendation](https://www.w3.org/TR/trace-context/#header-name) as far as I understand as we should be sending it as lowercase. 

`TraceContext` seems to have trouble consuming the header as uppercase but that seems like a separate issue.

In this PR I've used the underlying type of `http.Header` to ensure headers are kept and retrieved as it

### CODE REVIEW

**Review by MrAlias — CHANGES_REQUESTED** [review]
This change will correctly set the appropriate header value on outgoing propagation, but it will also incorrectly drop support for non-lowercase header values.

From the [W3C specification](https://www.w3.org/TR/trace-context/#header-name):

> In order to increase interoperability across multiple protocols and encourage successful integration, by default vendors SHOULD keep the header name lowercase.

This change would correct our behavior to comply with this directive. However ...

> Vendors MUST expect the header name in any case (upper, lower, mixed), and SHOULD send the header name in lowercase.

With this change the trace context propagator will only extract header values that are lowercase. Upper and mixed case header names will no longer be supported and we will stop complying with a required directive of the specification.

These changes need to include a fix that will continue to allow the trace context propagator to support the W3C specification.

**Comment by robertmarsal** [comment]
Thank you for the thorough review @MrAlias. I've incorporated your suggestions and had a go at resolving the issue of consuming the headers in any case by lowercasing them when we set them as this is the format for both of the supported headers. I've also expanded the test cases to check for this. Can you please have another look when you get a chance?

**Comment by MrAlias** [comment]
> Edit: please hold on the review as I discovered some issues with this approach.

Sounds good.

It's not required, but feel free to change this to a draft to help signal this. Otherwise I'll wait for the `WIP` prefix to change before I take another look.

Thanks for the diligence and the contribution :smile:

**Comment by robertmarsal** [comment]
I think I've actually hit a stumbling block with this at the moment. Because the underlying type of `HeaderCarrier` is `http.Header` users can build the `HeaderCarrier` using the native `http.Header` `Set` and that bypasses my attempts to avoid normalising like in this example from the tests:

```
req, _ := http.NewRequest("GET", "http://example.com", nil)
req.Header.Set("baggage", tt.header)

expected := tt.has.Baggage(t)
ctx := baggage.ContextWithBaggage(context.Background(), expected)
ctx = prop.Extract(ctx, propagation.HeaderCarrier(req.Header))
```

While I can fix this by checking for both the normalised and non normalised version when retrieving if the user does something like this:
```
req, _ := http.NewRequest("GET", "http://example.com", nil)
ctx := baggage.ContextWithBaggage(context.Background(), tt.mems.Baggage(t))
propagator.Inject(ctx, propagation.HeaderCarrier(req.Header))

got := strings.Split(req.Header.Get("baggage"), ",")
```
then the value will not be found because in the map it will be stored in lowercase.

It feels like this can't be accomplished with the current implementation and I don't think it's worth making big changes just for this. Unless there's something I am missing I think we can close this for now. Thank you

**Comment by robertmarsal** [comment]
@MrAlias to add more context to this, the way I discovered this was using a library that will receive the traceparent info from different sources like HTTP headers or other types of headers for example Kafka headers. This info was then converted to a text map and then extracted into a context. However this failed when the source was HTTP headers due to Go canonicalising the headers and `TextMap` being case-sensitive.

Basically the `TextMapCarrier` and `HeaderCarrier` implement the same interface however behave in different way and are not compatible with each other. Based on my research above there's no easy fix for this without changing the API but maybe we can improve the docs in this regard?

Thanks

**Comment by MrAlias** [comment]
Yeah, this sounds a bit tricky. My initial thought is to add an additional method to the `HeaderCarrier`, but I don't think this is necessarily the best approach. I'll have to think a bit about it.

@Aneurysm9 thoughts?

File: propagation/propagation.go [hunk 1]
```diff
@@ -69,14 +70,31 @@ func (c MapCarrier) Keys() []string {
...
-// Set stores the key-value pair.
+// Set sets the header entries associated with key to the single element value.
+// It replaces any existing values associated with key.
+//
+// The key is case-sensitive, it is used exactly as passed.
 func (hc HeaderCarrier) Set(key string, value string) {
-	http.Header(hc).Set(key, value)
+	// Use the underlying map to set the value to avoid normalising the key and
+	// thereby changing its value.
+	http.Header(hc)[strings.ToLower(key)] = []string{value}
```
Comment (Aneurysm9) [inline propagation/propagation.go:97]: Does this also need to check whether `hc == nil`?

I'm confused about the comment that says the value is set in the map directly to avoid normalization, but the key is passed through `strings.ToLower()` before use.

Comment (robertmarsal) [reply propagation/propagation.go:97]: the code and the comment might have drifted as I tried different options, however the conclusion I got to was that it was not possible with the current API to achieve what I wanted because of the reasons mentioned here: https://github.com/open-telemetry/opentelemetry-go/pull/3049#issuecomment-1204474240 I think this can be closed as it's not a valid solution at the moment
